File: quota_butler/planner.py
"""Deterministic V3 tomorrow-plan calculator."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Iterable, List, Mapping, Sequence, Tuple

from .providers.base import Usage
from .schedule_flow import PlanRequest, normalize_hhmm

DEFAULT_WINDOW_SECONDS = 5 * 3600
SUPPORTED_AGENTS = ("cc", "codex")
AGENT_LABELS = {"cc": "Claude Code", "codex": "Codex"}


@dataclass(frozen=True)
class PlanEvent:
    agent: str
    kind: str
    at: datetime
    purpose: str

    @property
    def note(self) -> str:
        return self.purpose


@dataclass(frozen=True)
class SchedulePlan:
    agents: Tuple[str, ...]
    work_start: datetime
    work_end: datetime
    events: Tuple[PlanEvent, ...]
    reason: str
    request: PlanRequest
    plan_version: int = 3

    @property
    def work_hours(self) -> float:
        return (self.work_end - self.work_start).total_seconds() / 3600


def build_plan(
    request: PlanRequest,
    available_usages: Mapping[str, Usage],
) -> SchedulePlan:
    start = _combine(request, request.work_start)
    end = _combine(request, request.work_end)
    if end <= start:
        raise ValueError("工作结束时间必须晚于开始时间")
    selected = _select_agents(
        request.agent_strategy,
        available_usages,
        (end - start).total_seconds() / 3600,
    )

    first_agent = selected[0]
    first_warmup = start - timedelta(hours=2, minutes=30)
    second_warmup = _real_reset_in_range(
        available_usages[first_agent],
        start,
        end,
    ) or first_warmup + timedelta(hours=5)
    events: List[PlanEvent] = [
        PlanEvent(first_agent, "warmup", first_warmup, "准备第一个窗口"),
        PlanEvent(first_agent, "warmup", second_warmup, "恢复后准备第二个窗口"),
    ]

    if len(selected) == 2:
        relay_at = max(start + timedelta(hours=5) - timedelta(minutes=10), start)
        if relay_at < end:
            events.append(
                PlanEvent(selected[1], "warmup", relay_at, "为长时间工作准备接力窗口")
            )
        reason = (
            f"前半段优先保持 {AGENT_LABELS[first_agent]} 连续工作，"
            f"后半段由 {AGENT_LABELS[selected[1]]} 接力。"
        )
    else:
        reason = (
            f"当前区间优先使用单 Agent：{AGENT_LABELS[first_agent]}，"
            "减少上下文切换。"
        )
    events.sort(key=lambda item: (item.at, item.agent))
    return SchedulePlan(
        agents=selected,
        work_start=start,
        work_end=end,
        events=tuple(events),
        reason=reason,
        request=request,
    )


def parse_agents(value: object) -> Tuple[str, ...]:
    if isinstance(value, str):
        raw = value.replace("，", ",").split(",")
    elif isinstance(value, Iterable):
        raw = list(value)
    else:
        raw = []
    agents: List[str] = []
    for item in raw:
        agent = _normalize_agent(str(item))
        if agent not in agents:
            agents.append(agent)
    return tuple(agents)


def _select_agents(
    strategy: str,
    usages: Mapping[str, Usage],
    work_hours: float,
) -> Tuple[str, ...]:
    available = tuple(agent for agent in SUPPORTED_AGENTS if agent in usages)
    if not available:
        raise ValueError("当前没有可用于规划的 Agent")
    if strategy in ("cc", "codex"):
        if strategy not in usages:
            raise ValueError(f"{AGENT_LABELS[strategy]} 当前不可用")
        return (strategy,)
    if strategy == "both":
        if len(available) < 2:
            raise ValueError("Claude Code + Codex 当前不能同时使用")
        return _rank_agents(available, usages)
    ranked = _rank_agents(available, usages)
    if work_hours > 5 and len(ranked) > 1:
        return ranked
    return (ranked[0],)


def _rank_agents(
    agents: Sequence[str],
    usages: Mapping[str, Usage],
) -> Tuple[str, ...]:
    return tuple(
        sorted(
            agents,
            key=lambda agent: (
                usages[agent].five_hour.utilization,
                SUPPORTED_AGENTS.index(agent),
            ),
        )
    )


def _real_reset_in_range(
    usage: Usage,
    start: datetime,
    end: datetime,
):
    reset = usage.five_hour.resets_at
    if reset is None:
        return None
    if reset.tzinfo is not None:
        reset = reset.astimezone().replace(tzinfo=None)
    if start <= reset <= end:
        return reset
    return None


def _combine(request: PlanRequest, hhmm: str) -> datetime:
    normalized = normalize_hhmm(hhmm)
    hour, minute = (int(part) for part in normalized.split(":"))
    return datetime.combine(
        request.target_date,
        datetime.min.time().replace(hour=hour, minute=minute),
    )


def _normalize_agent(value: str) -> str:
    key = value.strip().lower()
    if key in ("claude", "claude-code", "claude code"):
        key = "cc"
    if key not in SUPPORTED_AGENTS:
        raise ValueError(f"unsupported scheduler agent: {value!r}")
    return key

File: src/planner.ts
// 确定性 V3 明日计划计算器：一次只规划一个 AI 工具，用两次预热最大化单工具窗口。

import type { PlanRequest } from './schedule_flow.js';
import { normalizeHHmm } from './schedule_flow.js';
import type { Usage } from './providers/index.js';

export const SUPPORTED_AGENTS = ['cc', 'codex'] as const;
export const AGENT_LABELS: Record<string, string> = { cc: 'Claude Code', codex: 'Codex' };

export interface PlanEvent {
  agent: string;
  kind: string;
  at: Date;
  purpose: string;
}

export interface SchedulePlan {
  agents: string[];
  workStart: Date;
  workEnd: Date;
  events: PlanEvent[];
  reason: string;
  request: PlanRequest;
  planVersion: number;
}

const HOUR = 3600000;

export function buildPlan(request: PlanRequest, availableUsages: Record<string, Usage>): SchedulePlan {
  const start = combine(request, request.workStart);
  let end = combine(request, request.workEnd);
  // 结束 ≤ 开始 → 视为次日（支持跨午夜/夜班）。预热点、relay 全用绝对时间，跨天自然成立。
  if (end.getTime() <= start.getTime()) end = new Date(end.getTime() + 24 * HOUR);
  const selected = selectAgents(request.agentStrategy, availableUsages);

  const firstAgent = selected[0]!;
  const firstWarmup = combine(request, request.firstWarmup);
  const secondWarmup = combine(request, request.secondWarmup);
  const sortedWarmups = [firstWarmup, secondWarmup].sort((a, b) => a.getTime() - b.getTime());
  if (end.getTime() <= sortedWarmups[1]!.getTime()) end = new Date(sortedWarmups[1]!.getTime() + 5 * HOUR);

  const events: PlanEvent[] = [
    { agent: firstAgent, kind: 'warmup', at: sortedWarmups[0]!, purpose: '准备第一个窗口' },
    { agent: firstAgent, kind: 'warmup', at: sortedWarmups[1]!, purpose: '恢复后准备第二个窗口' },
  ];

  const reason = `当前计划只使用 ${AGENT_LABELS[firstAgent]}，用两次预热最大化单一工具的可用窗口。`;

  events.sort((a, b) => a.at.getTime() - b.at.getTime() || a.agent.localeCompare(b.agent));
  return { agents: selected, workStart: start, workEnd: end, events, reason, request, planVersion: 3 };
}

export function parseAgents(value: unknown): string[] {
  let raw: unknown[];
  if (typeof value === 'string') raw = value.replace(/，/g, ',').split(',');
  else if (Array.isArray(value)) raw = value;
  else raw = [];
  const agents: string[] = [];
  for (const item of raw) {
    const agent = normalizeAgent(String(item));
    if (!agents.includes(agent)) agents.push(agent);
  }
  return agents;
}

function selectAgents(strategy: string, usages: Record<string, Usage>): string[] {
  const available = (SUPPORTED_AGENTS as readonly string[]).filter((a) => a in usages);
  if (available.length === 0) throw new Error('当前没有可用于规划的 Agent');
  if (strategy === 'cc' || strategy === 'codex') {
    if (!(strategy in usages)) throw new Error(`${AGENT_LABELS[strategy]} 当前不可用`);
    return [strategy];
  }
  if (strategy === 'both') {
    throw new Error('当前流程一次只编排一个 AI 工具');
  }
  const ranked = rankAgents(available, usages);
  return [ranked[0]!];
}

function rankAgents(agents: string[], usages: Record<string, Usage>): string[] {
  // 周额度（木桶上限）剩余多的优先，其次 5 小时剩余多的优先。
  return [...agents].sort((a, b) => {
    const wa = weeklyRemaining(usages[a]!);
    const wb = weeklyRemaining(usages[b]!);
    if (wa !== wb) return wb - wa;
    // 规划候选必有 5h 窗口（usableForPlanning 已保证），仍做空值兜底。
    const fa = usages[a]!.fiveHour ? 100 - usages[a]!.fiveHour!.utilization : 0;
    const fb = usages[b]!.fiveHour ? 100 - usages[b]!.fiveHour!.utilization : 0;
    if (fa !== fb) return fb - fa;
    return (SUPPORTED_AGENTS as readonly string[]).indexOf(a) - (SUPPORTED_AGENTS as readonly string[]).indexOf(b);
  });
}

function weeklyRemaining(usage: Usage): number {
  return usage.sevenDay ? 100 - usage.sevenDay.utilization : 100;
}

function combine(request: PlanRequest, hhmm: string): Date {
  const n = normalizeHHmm(hhmm);
  const [h, m] = n.split(':').map(Number) as [number, number];
  const [y, mo, d] = request.targetDate.split('-').map(Number) as [number, number, number];
  return new Date(y, mo - 1, d, h, m, 0, 0);
}

function normalizeAgent(value: string): string {
  let key = value.trim().toLowerCase();
  if (['claude', 'claude-code', 'claude code'].includes(key)) key = 'cc';
  if (!(SUPPORTED_AGENTS as readonly string[]).includes(key)) {
    throw new Error(`unsupported scheduler agent: ${value}`);
  }
  return key;
}
